File: AOC_2023/Day 21/21.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
#include <set>

using namespace std;

using ll = long long;

class Solution
{
    public:
    int n, m;
    vector<string> grid;
    pair<int,int> start;
    pair<int,int> dirs[4] = {{0,1},{1,0},{0,-1},{-1,0}};
    void init(ifstream& input){
        string line;
        while (getline(input, line))
            grid.push_back(line);
        n = grid.size();
        m = grid[0].size();
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[i].size(); j++){
                if(grid[i][j] == 'S'){
                    start = {i,j};
                    break;
                }
            }
        }
    }

    bool isValid(pair<int,int> pos){
        while(pos.first < 0) pos.first += n;
        while(pos.second < 0) pos.second += m;
        pos.first %= n;
        pos.second %= m;
        return grid[pos.first][pos.second] == '.';
    }

    void part1(){
        queue<pair<int,int>> q;
        q.push(start);
        int steps = 0;
        while(steps < 64 && !q.empty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                pair<int,int> curr = q.front();
                q.pop();
                grid[curr.first][curr.second] = '.';
                for(int j = 0; j < 4; j++){
                    pair<int,int> next = {curr.first + dirs[j].first, curr.second + dirs[j].second};
                    if(isValid(next)){
                        grid[next.first][next.second] = 'O';
                        q.push(next);
                    }
                }
            }
            steps++;
        }
        cout<<"Part 1: "<< q.size() << "\n";
    }

    ll function(ll n, vector<int> temp){
        ll  y0 = temp[0], y1 = temp[1], y2 = temp[2];
        ll a = (y2 - 2*y1 + y0)/2;
        ll b = y2 - y1 - a;
        ll c = y1;
        ll result = a*n*n + b*n + c;
        return result;
    }    

    void part2(){
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[i].size(); j++){
                if(grid[i][j] == 'O'){
                    grid[i][j] = '.';
                }
            }
        }
        grid[start.first][start.second] = '.';
        vector<int> temp;
        int steps = 0;
        set<pair<int,int>> prevVisited, currVisited;
        currVisited.insert(start);
        while(true){
            int size = currVisited.size();
            if(steps%n == n/2) temp.push_back(size);
            if(temp.size() == 3) break;
            prevVisited.clear();
            for(auto it : currVisited){
                for(int j = 0; j < 4; j++){
                    pair<int,int> next = {it.first + dirs[j].first, it.second + dirs[j].second};
                    if(isValid(next)){
                        prevVisited.insert(next);
                    }
                }
            }
            currVisited = prevVisited;
            steps++;
        }
        cout<<"Part 2: "<< function(ll(26501365/n), temp) << "\n";
    }
};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2023/Day 21/21.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type pair struct {
	first, second int
}

type Solution struct {
	n, m   int
	grid   []string
	start  pair
	dirs   []pair
	visited map[pair]bool
}

func (s *Solution) init(input *os.File) {
	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		s.grid = append(s.grid, scanner.Text())
	}
	s.n = len(s.grid)
	s.m = len(s.grid[0])
	s.dirs = []pair{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
	s.visited = make(map[pair]bool)
	for i := 0; i < s.n; i++ {
		for j := 0; j < s.m; j++ {
			if s.grid[i][j] == 'S' {
				s.start = pair{i, j}
				break
			}
		}
	}
}

func (s *Solution) isValid(pos pair) bool {
	for pos.first < 0 {
		pos.first += s.n
	}
	for pos.second < 0 {
		pos.second += s.m
	}
	pos.first %= s.n
	pos.second %= s.m
	return s.grid[pos.first][pos.second] == '.'
}

func (s *Solution) part1() {
	q := []pair{s.start}
	steps := 0

	for steps < 64 && len(q) > 0 {
		size := len(q)
		for i := 0; i < size; i++ {
			curr := q[0]
			q = q[1:]
			s.grid[curr.first] = s.grid[curr.first][:curr.second] + "." + s.grid[curr.first][curr.second+1:]

			for _, dir := range s.dirs {
				next := pair{curr.first + dir.first, curr.second + dir.second}
				if s.isValid(next) && s.grid[next.first][next.second] == '.' {
					s.grid[next.first] = s.grid[next.first][:next.second] + "O" + s.grid[next.first][next.second+1:]
					q = append(q, next)
				}
			}
		}
		steps++
	}
	fmt.Println("Part 1:", len(q))
}

func function(n int64, temp []int) int64 {
	y0, y1, y2 := temp[0], temp[1], temp[2]
	a := (int64(y2) - 2*int64(y1) + int64(y0)) / 2
	b := int64(y2) - int64(y1) - a
	c := int64(y1)
	result := a*n*n + b*n + c
	return result
}

func (s *Solution) part2() {
	for i := 0; i < s.n; i++ {
		for j := 0; j < s.m; j++ {
			if s.grid[i][j] == 'O' {
				s.grid[i] = s.grid[i][:j] + "." + s.grid[i][j+1:]
			}
		}
	}
	s.grid[s.start.first] = s.grid[s.start.first][:s.start.second] + "." + s.grid[s.start.first][s.start.second+1:]
	temp := make([]int, 0)
	steps := 0
	prevVisited, currVisited := make(map[pair]bool), make(map[pair]bool)
	currVisited[s.start] = true

	for {
		size := len(currVisited)
		if steps%s.n == s.n/2 {
			temp = append(temp, size)
		}
		if len(temp) == 3 {
			break
		}
		prevVisited = make(map[pair]bool)
		for it := range currVisited {
			for _, dir := range s.dirs {
				next := pair{it.first + dir.first, it.second + dir.second}
				if s.isValid(next) {
					prevVisited[next] = true
				}
			}
		}
		currVisited = prevVisited
		steps++
	}
	fmt.Println("Part 2:", function(int64(26501365/s.n), temp))
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}
	defer file.Close()
	s := Solution{}
	s.init(file)
	s.part1()
	s.part2()
}
